### PR TITLE
Feature/routing

### PR DESCRIPTION
**Pull request by kuzmavictor**
At the moment application contains implemented registration and authorization forms with validation services. 
Added new feature:

- route service;
- preview image for avatar on the registration form.

### CODE REVIEW

File: app/dto/error/authorization/index.js [hunk 1]
```diff
@@ -0,0 +1,30 @@
+/**
+ * This class is DTO, that contains errors validation form.
+ */
+export class ErrorAuthoritationDTO {
```
Comment (dkalinin) [inline app/dto/error/authorization/index.js]: Authori-z-ation

Comment (dkalinin) [reply app/dto/error/authorization/index.js]: AuthorizationErrorDto.

File: app/page/index.js [hunk 2]
```diff
@@ -0,0 +1,8 @@
+export class Page {
```
Comment (dkalinin) [inline app/page/index.js]: This class doesn't add any functionality.

File: app/page/registration/index.js [hunk 3]
```diff
@@ -0,0 +1,11 @@
+import {Page} from '../index';
+
+export class RegistrationPage extends Page {
+  constructor(parentRoot) {
+    super(parentRoot);
+  }
+
+  render() {
+
+  }
```
Comment (dkalinin) [inline app/page/registration/index.js]: Why is it empty?

File: app/service/router/index.js [hunk 4]
```diff
@@ -0,0 +1,78 @@
+/**
+ *
+ */
```
Comment (dkalinin) [inline app/service/router/index.js]: No empty JS doc please.

File: app/service/router/index.js [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+    window.addEventListener('hashchange', () => this._handleRouteChanged());
+  }
+
+  /**
+   *
+   * @private
+   * @return {Object}
+   */
+  _handleRouteChanged() {
+    console.log(this._routingConfig);
```
Comment (dkalinin) [inline app/service/router/index.js]: It's not needed.

**Comment by kuzmavictor** [comment]
@dkalinin PTAL.

**Comment by kuzmavictor** [comment]
@dkalinin PTAL.

image

File: app/service/validation/index.js [hunk 6]
```diff
@@ -21,6 +22,16 @@ export class ValidationService {
       || typeof errorRegistrationDTO.confirmedPasswordError !== 'undefined';
   }
 
+  /**
+   * Returns flag of errors. If at authorization there is at least one error will be return true.
```
Comment (dkalinin) [inline app/service/validation/index.js]: Checks if there's any authorization error.

File: app/service/validation/index.js [hunk 7]
```diff
@@ -21,6 +22,16 @@ export class ValidationService {
...
+   *
+   * @param {Object} authorizationErrorDTO - a DTO that contains errors of registration
+   * @return {boolean} the flag errors.
+   */
+  isErrorAuthorization(authorizationErrorDTO) {
```
Comment (dkalinin) [inline app/service/validation/index.js]: Let's move it to a getter of ErrorDTO. LIke `get hasError()`.

File: app/service/validation/index.js [hunk 8]
```diff
@@ -21,6 +22,16 @@ export class ValidationService {
...
+    return typeof authorizationErrorDTO.loginError !== 'undefined'
+      || typeof authorizationErrorDTO.passwordError !== 'undefined';
+  }
 
   /**
    *Validate data that contains inside {@link RegistrationDTO}.
```
Comment (dkalinin) [inline app/service/validation/index.js]: There should be a space between * and Validate.

File: app/service/validation/index.js [hunk 9]
```diff
@@ -33,31 +44,46 @@ export class ValidationService {
...
     return registrationErrorDto;
   }
 
+  /**
+   *  Validate data that contains inside {@link AuthorizationDTO}.
+   *
+   * @param {Object} authorizationDTO -an DTO that contains data from authorization form
+   * @return {Object} - the DTO that will be contain error message if errors will  be detect.
+   */
+  validateAuthorizationForm(authorizationDTO) {
```
Comment (dkalinin) [inline app/service/validation/index.js]: Let's have separate services for Login and Registration validations.

File: app/view/component/button/index.js [hunk 10]
```diff
@@ -39,23 +39,13 @@ export class Button extends Component {
...
-  }
-
   /**
    * Adds event listener to button component.
    *
    * @private
    */
   _addEventHandler() {
-    this._button.addEventListener( 'click', (event) => this.callback( event ) );
+    this.button.addEventListener( 'click', (event) => this.callback( event ) );
```
Comment (dkalinin) [inline app/view/component/button/index.js]: If there's no callback passed then error will be thrown.

File: app/view/component/form/authorization/index.js [hunk 11]
```diff
@@ -0,0 +1,162 @@
...
+  /**
+   * Handlers action 'click'  of button component. In fact  it is the body
+   * of callback function (action).
+   *
+   * @param {Function} event - an callback function for button click
+   * @private
+   */
+  _buttonOnClickHandler(event) {
+    const loginData = this.login.getInputValue().value;
+    const passwordData = this.password.getInputValue().value;
```
Comment (dkalinin) [inline app/view/component/form/authorization/index.js]: getInputValue().value looks weird. A getter of input value should return proper value.

File: app/view/component/form/authorization/index.js [hunk 12]
```diff
@@ -0,0 +1,162 @@
...
+   * may be during registration process
+   */
+  reRenderAuthorizationForm(formErrorDTO) {
+    const loginError = formErrorDTO.loginError;
+    const passwordError = formErrorDTO.passwordError;
+    if (typeof loginError !== 'undefined') {
+      this.login.textError = loginError;
+      this.login.reRender();
+    }
+    if (typeof passwordError !== 'undefined') {
```
Comment (dkalinin) [inline app/view/component/form/authorization/index.js]: What if we don't check for undefined?

File: app/view/component/form/authorization/index.js [hunk 13]
```diff
@@ -0,0 +1,162 @@
...
+   * @return {Object} - the object that contains configuration for {@link FormField}
+   * (inputLoginId, labelText, placeHolderText, loginError, inputType)
+   * @private
+   */
+  _configLoginFiledOfAuthorizationForm() {
+    const inputLoginId = {inputId: 'input-auth-login-id'};
+    const labelText = {labelText: 'Login'};
+    const placeHolderText = {placeHolderText: 'Input login'};
+    const errorMessage = {errorMessage: ''};
+    const inputType = {inputType: 'text'};
```
Comment (dkalinin) [inline app/view/component/form/authorization/index.js]: All these local variables are not needed.

File: app/view/component/form/authorization/index.js [hunk 14]
```diff
@@ -0,0 +1,162 @@
...
+    return this.configLoginField = Object.assign(inputLoginId, labelText,
```
Comment (dkalinin) [inline app/view/component/form/authorization/index.js]: Is `configLoginField` field used anywhere?

File: app/view/component/form/authorization/index.js (hunk 13 above)
Comment (dkalinin) [reply app/view/component/form/authorization/index.js]: Same comments for other config methods.

File: app/view/component/form/authorization/index.js (hunk 14 above)
Comment (dkalinin) [reply app/view/component/form/authorization/index.js]: Same comments as for other config methods.

File: app/view/component/form/authorization/index.js (hunk 12 above)
Comment (kuzmavictor) [reply app/view/component/form/authorization/index.js]: The `undefined` is a default value of fields inside some  DTOs that contain validation errors.  If the field is not equal to `undefined`  then it contains an error message.  This means that in the form field that contains the error, the component that displays this error will be rerendered.

**Comment by kuzmavictor** [comment]
@dkalinin fixed, PTAL.

File: app/view/component/form/authorization/index.js (hunk 11 above)
Comment (dkalinin) [reply app/view/component/form/authorization/index.js]: @kuzmavictor don't mark unfixed comments as resolved.

File: app/view/component/form/authorization/index.js (hunk 14 above)
Comment (dkalinin) [reply app/view/component/form/authorization/index.js]: @kuzmavictor don't mark unfixed comments as resolved.

File: app/view/component/form/authorization/index.js (hunk 13 above)
Comment (dkalinin) [reply app/view/component/form/authorization/index.js]: @kuzmavictor don't mark unfixed comments as resolved.

File: app/view/component/form/authorization/index.js [hunk 15]
```diff
@@ -0,0 +1,151 @@
...
+
+  /**
+   * Returns configuration  for {@link FormField} component (a field to input password).
+   *
+   * @return {Object} - the object that contains configuration for {@link FormField}
+   * (inputLoginId, labelText, placeHolderText, errorMessage, inputType)
+   * @private
+   */
+  static _configPasswordFieldOfAuthorizationForm() {
+    return Object.assign({inputId: 'input-auth-psw-id'}, {labelText: 'Password: '},
```
Comment (dkalinin) [inline app/view/component/form/authorization/index.js]: Why do we need assign here?

File: app/view/component/form/authorization/index.js [hunk 16]
```diff
@@ -0,0 +1,151 @@
...
+
+  /**
+   * Returns configuration  for {@link FormField} component (a field to input  login).
+   *
+   * @return {Object} - the object that contains configuration for {@link FormField}
+   * (inputLoginId, labelText, placeHolderText, loginError, inputType)
+   * @private
+   */
+  static _configLoginFiledOfAuthorizationForm() {
+    return Object.assign({inputId: 'input-auth-login-id'}, {labelText: 'Login'},
```
Comment (dkalinin) [inline app/view/component/form/authorization/index.js]: Why do we need assign here?

File: app/view/component/form/authorization/index.js (hunk 15 above)
Comment (kuzmavictor) [reply app/view/component/form/authorization/index.js]: This way allows making the constructor of components on the form less redundant.

File: app/view/component/form/authorization/index.js (hunk 16 above)
Comment (kuzmavictor) [reply app/view/component/form/authorization/index.js]: This way allows making the constructor of components on the form less redundant.

**Pull request merged**